src/simulator/simulator.cpp

```diff
@@ -126,48 +139,106 @@ void Simulator::init(ros::NodeHandle& n) {
   }
-  motion_model_->SetPose(cur_loc_);
-  initSimulatorVizMarkers();
+
+  initSimulatorVizMarkers(); // TODO: need to add a vis marker for every robot
   drawMap();
 
-  initSubscriber = n.subscribe(
-      "/initialpose", 1, &Simulator::InitalLocationCallback, this);
-  odometryTwistPublisher = n.advertise<nav_msgs::Odometry>("/odom",1);
-  laserPublisher = n.advertise<sensor_msgs::LaserScan>(CONFIG_laser_topic, 1);
-  viz_laser_publisher_ = n.advertise<sensor_msgs::LaserScan>("/scan", 1);
-  mapLinesPublisher = n.advertise<visualization_msgs::Marker>(
-      "/simulator_visualization", 6);
+  // Initialize topics for each robot
+  for(int i = 0; i < robot_number_; i++){
+    string topic_prefix = topic_prefixs_[i];
+    printf("[%d] topic_prefix: %s\n", i, topic_prefix.c_str());
+    // TODO: Change this hardcoding to something nicer
+        odometryTwistPublishers.push_back(
+        n.advertise<nav_msgs::Odometry>(topic_prefix + "/odom", 1));
+    laserPublishers.push_back(n.advertise<sensor_msgs::LaserScan>(
+        topic_prefix + CONFIG_laser_topic, 1));
+    viz_laser_publishers_.push_back(n.advertise<sensor_msgs::LaserScan>(
+        topic_prefix + "/scan", 1));
+    if (FLAGS_localize) {
+      localizationPublishers.push_back(n.advertise<ut_multirobot_sim::Localization2DMsg>(
+          topic_prefix + "/localization", 1));
+    }
+  }
+  // initialize hard coded initial pose
+  for (int i = 0; i < robot_number_; i++) {
+    if (i == 7) {
+      initSubscribers.push_back(
+          n.subscribe(topic_prefixs_[7] + "/initialpose", 1,
+                      &Simulator::InitalLocationCallbackCar7, this));
+    } else if (i == 6) {
+      initSubscribers.push_back(
+          n.subscribe(topic_prefixs_[6] + "/initialpose", 1,
+                      &Simulator::InitalLocationCallbackCar6, this));
+    } else if (i == 5) {
+      initSubscribers.push_back(
+          n.subscribe(topic_prefixs_[5] + "/initialpose", 1,
+                      &Simulator::InitalLocationCallbackCar5, this));
+    } else if (i == 4) {
+      initSubscribers.push_back(
+          n.subscribe(topic_prefixs_[4] + "/initialpose", 1,
+                      &Simulator::InitalLocationCallbackCar4, this));
+    } else if (i == 3) {
+      initSubscribers.push_back(
+          n.subscribe(topic_prefixs_[3] + "/initialpose", 1,
+                      &Simulator::InitalLocationCallbackCar3, this));
+    } else if (i == 2) {
+      initSubscribers.push_back(
+          n.subscribe(topic_prefixs_[2] + "/initialpose", 1,
+                      &Simulator::InitalLocationCallbackCar2, this));
+    } else if (i == 1) {
+      initSubscribers.push_back(
+          n.subscribe(topic_prefixs_[1] + "/initialpose", 1,
+                      &Simulator::InitalLocationCallbackCar1, this));
+    } else if (i == 0) {
+      initSubscribers.push_back(
+          n.subscribe(topic_prefixs_[0] + "/initialpose", 1,
+                      &Simulator::InitalLocationCallbackCar0, this));
```

Comment:
I am not sure if we should hard code like this. Need other's comments

Reply:
This relates to the hardcoded unwrapped function below. Assuming that different function names are used I believe that this is the most concise option. Of course, if we can find a workaround of using only one singular function instead that will be great.

Reply:
We should have a conversation about this. Whatever is happening I'm convinced that we don't want to be doing this, but I'm not 100% clear on why the choice seems to be between this and boost higher order functions. It seems to me like you would just need to put the robot_number in the initial location message and this would be handled by one function.

If this is done this way to automatically work with RVIZ, we can just not do that I think.